File: pbmx_ffi/java/io/rmf/pbmx/BlockBuilder.java
package io.rmf.pbmx;

import static io.rmf.pbmx.Util.toMaskArray;
import static io.rmf.pbmx.Util.toShareArray;
import static io.rmf.pbmx.Util.toIdArray;
import static io.rmf.pbmx.Util.toMaskPointerArray;
import static io.rmf.pbmx.Util.toSharePointerArray;
import io.rmf.pbmx.ffi.LibPbmx;
import io.rmf.pbmx.ffi.RawMask;
import io.rmf.pbmx.ffi.RawShare;
import io.rmf.pbmx.ffi.RawFingerprint;
import io.rmf.pbmx.payloads.Payload;
import java.util.Collection;
import java.nio.ByteBuffer;
import com.sun.jna.Pointer;

public final class BlockBuilder {
    BlockBuilder(Pointer pbmx, Pointer handle) {
        this.pbmx = pbmx;
        this.handle = handle;
    }

    Block build() {
        if(deleted) {
            throw new IllegalArgumentException();
        }
        this.deleted = true;
        return new Block(LibPbmx.INSTANCE.pbmx_build_block(this.pbmx, this.handle));
    }

    public void add(Payload payload) {
        payload.addTo(this);
    }

    public void publishKey(String name, PublicKey key) {
        if(deleted) {
            throw new IllegalArgumentException();
        }
        int r = LibPbmx.INSTANCE.pbmx_publish_key_payload(this.handle, name, key.handle);
        assert r != 0;
    }

    public void openStack(Collection<Mask> stack) {
        if(deleted) {
            throw new IllegalArgumentException();
        }
        RawMask[] masks = toMaskArray(stack);
        int r = LibPbmx.INSTANCE.pbmx_open_stack_payload(this.handle, masks, masks.length);
        assert r != 0;
    }

    public void maskStack(Id id, Collection<Mask> stack, Collection<Mask.Proof> proofs) {
        if(deleted) {
            throw new IllegalArgumentException();
        }
        RawMask[] masks = toMaskArray(stack);
        Pointer[] proofPtrs = toMaskPointerArray(proofs);
        int r = LibPbmx.INSTANCE.pbmx_mask_stack_payload(this.handle, id.raw.val(), masks, masks.length, proofPtrs);
        assert r != 0;
    }

    public void shuffleStack(Id id, Collection<Mask> stack, ShuffleProof proof) {
        if(deleted) {
            throw new IllegalArgumentException();
        }
        RawMask[] masks = toMaskArray(stack);
        int r = LibPbmx.INSTANCE.pbmx_shuffle_stack_payload(this.handle, id.raw.val(), masks, masks.length, proof.handle);
        assert r != 0;
    }

    public void shiftStack(Id id, Collection<Mask> stack, ShiftProof proof) {
        if(deleted) {
            throw new IllegalArgumentException();
        }
        RawMask[] masks = toMaskArray(stack);
        int r = LibPbmx.INSTANCE.pbmx_shift_stack_payload(this.handle, id.raw.val(), masks, masks.length, proof.handle);
        assert r != 0;
    }

    public void nameStack(Id id, String name) {
        if(deleted) {
            throw new IllegalArgumentException();
        }
        int r = LibPbmx.INSTANCE.pbmx_name_stack_payload(this.handle, id.raw.val(), name);
        assert r != 0;
    }

    public void takeStack(Id id1, long[] indices, Id id2) {
        if(deleted) {
            throw new IllegalArgumentException();
        }
        int r = LibPbmx.INSTANCE.pbmx_take_stack_payload(this.handle, id1.raw.val(), indices, indices.length, id2.raw.val());
        assert r != 0;
    }

    public void pileStacks(Collection<Id> ids, Id id) {
        if(deleted) {
            throw new IllegalArgumentException();
        }
        RawFingerprint[] fps = toIdArray(ids);
        int r = LibPbmx.INSTANCE.pbmx_pile_stacks_payload(this.handle, fps, fps.length, id.raw.val());
        assert r != 0;
    }

    public void insertStack(Id id1, Id id2, Collection<Mask> stack, InsertProof proof) {
        if(deleted) {
            throw new IllegalArgumentException();
        }
        RawMask[] masks = toMaskArray(stack);
        int r = LibPbmx.INSTANCE.pbmx_insert_stack_payload(this.handle, id1.raw.val(), id2.raw.val(), masks, masks.length, proof.handle);
        assert r != 0;
    }

    public void publishShares(Id id, Collection<Share> shares, Collection<Share.Proof> proofs) {
        if(deleted) {
            throw new IllegalArgumentException();
        }
        RawShare[] raws = toShareArray(shares);
        Pointer[] proofPtrs = toSharePointerArray(proofs);
        int r = LibPbmx.INSTANCE.pbmx_publish_shares_payload(this.handle, id.raw.val(), raws, raws.length, proofPtrs);
        assert r != 0;
    }

    public void randomSpec(String name, String spec) {
        if(deleted) {
            throw new IllegalArgumentException();
        }
        int r = LibPbmx.INSTANCE.pbmx_random_spec_payload(this.handle, name, spec);
        assert r != 0;
    }

    public void randomEntropy(String name, Mask entropy) {
        if(deleted) {
            throw new IllegalArgumentException();
        }
        int r = LibPbmx.INSTANCE.pbmx_random_entropy_payload(this.handle, name, entropy.raw.val());
        assert r != 0;
    }

    public void randomReveal(String name, Share share, Share.Proof proof) {
        if(deleted) {
            throw new IllegalArgumentException();
        }
        int r = LibPbmx.INSTANCE.pbmx_random_reveal_payload(this.handle, name, share.raw.val(), proof.handle);
        assert r != 0;
    }

    public void bytes(ByteBuffer buf) {
        int r = LibPbmx.INSTANCE.pbmx_bytes_payload(this.handle, buf, buf.remaining());
        assert r != 0;
    }

    @Override
    protected void finalize() {
        if(!deleted) {
            LibPbmx.INSTANCE.pbmx_delete_block_builder(this.handle);
        }
    }

    Pointer pbmx;
    Pointer handle;
    boolean deleted = false;
}

File: pbmx_droid/app/src/main/java/io/rmf/pbmx/BlockBuilder.kt
package io.rmf.pbmx

import com.sun.jna.Pointer
import io.rmf.pbmx.ffi.FFI
import java.nio.ByteBuffer

class BlockBuilder internal constructor(internal var pbmx: Pointer, internal var handle: Pointer) {
    private var deleted = false

    internal fun build(): Block {
        if (deleted) {
            throw IllegalArgumentException()
        }
        this.deleted = true
        return Block(FFI.pbmx_build_block(this.pbmx, this.handle))
    }

    fun publishKey(name: String, key: PublicKey) {
        if (deleted) {
            throw IllegalArgumentException()
        }
        val r = FFI.pbmx_publish_key_payload(this.handle, name, key.handle)
        assert(r != 0)
    }

    fun openStack(stack: Collection<Mask>) {
        if (deleted) {
            throw IllegalArgumentException()
        }
        val masks = toMaskArray(stack)
        val r = FFI.pbmx_open_stack_payload(this.handle, masks, masks.size.toLong())
        assert(r != 0)
    }

    fun maskStack(id: Id, stack: Collection<Mask>, proofs: Collection<Mask.Proof>) {
        if (deleted) {
            throw IllegalArgumentException()
        }
        val masks = toMaskArray(stack)
        val proofPtrs = toMaskPointerArray(proofs)
        val r =
            FFI.pbmx_mask_stack_payload(this.handle, id.raw.value(), masks, masks.size.toLong(), proofPtrs)
        assert(r != 0)
    }

    fun shuffleStack(id: Id, stack: Collection<Mask>, proof: ShuffleProof) {
        if (deleted) {
            throw IllegalArgumentException()
        }
        val masks = toMaskArray(stack)
        val r =
            FFI.pbmx_shuffle_stack_payload(
                this.handle,
                id.raw.value(),
                masks,
                masks.size.toLong(),
                proof.handle
            )
        assert(r != 0)
    }

    fun shiftStack(id: Id, stack: Collection<Mask>, proof: ShiftProof) {
        if (deleted) {
            throw IllegalArgumentException()
        }
        val masks = toMaskArray(stack)
        val r = FFI.pbmx_shift_stack_payload(
            this.handle,
            id.raw.value(),
            masks,
            masks.size.toLong(),
            proof.handle
        )
        assert(r != 0)
    }

    fun nameStack(id: Id, name: String) {
        if (deleted) {
            throw IllegalArgumentException()
        }
        val r = FFI.pbmx_name_stack_payload(this.handle, id.raw.value(), name)
        assert(r != 0)
    }

    fun takeStack(id1: Id, indices: LongArray, id2: Id) {
        if (deleted) {
            throw IllegalArgumentException()
        }
        val r = FFI.pbmx_take_stack_payload(
            this.handle,
            id1.raw.value(),
            indices,
            indices.size.toLong(),
            id2.raw.value()
        )
        assert(r != 0)
    }

    fun pileStacks(ids: Collection<Id>, id: Id) {
        if (deleted) {
            throw IllegalArgumentException()
        }
        val fps = toIdArray(ids)
        val r = FFI.pbmx_pile_stacks_payload(this.handle, fps, fps.size.toLong(), id.raw.value())
        assert(r != 0)
    }

    fun publishShares(id: Id, shares: Collection<Share>, proofs: Collection<Share.Proof>) {
        if (deleted) {
            throw IllegalArgumentException()
        }
        val raws = toShareArray(shares)
        val proofPtrs = toSharePointerArray(proofs)
        val r = FFI.pbmx_publish_shares_payload(
            this.handle, id.raw.value(), raws,
            raws.size.toLong(), proofPtrs
        )
        assert(r != 0)
    }

    fun randomSpec(name: String, spec: String) {
        if (deleted) {
            throw IllegalArgumentException()
        }
        val r = FFI.pbmx_random_spec_payload(this.handle, name, spec)
        assert(r != 0)
    }

    fun randomEntropy(name: String, entropy: Mask) {
        if (deleted) {
            throw IllegalArgumentException()
        }
        val r = FFI.pbmx_random_entropy_payload(this.handle, name, entropy.raw.value())
        assert(r != 0)
    }

    fun randomReveal(name: String, share: Share, proof: Share.Proof) {
        if (deleted) {
            throw IllegalArgumentException()
        }
        val r = FFI.pbmx_random_reveal_payload(this.handle, name, share.raw.value(), proof.handle)
        assert(r != 0)
    }

    fun text(str: String) {
        if (deleted) {
            throw IllegalArgumentException()
        }
        val r = FFI.pbmx_text_payload(this.handle, str)
        assert(r != 0)
    }

    fun bytes(buf: ByteBuffer) {
        if (deleted) {
            throw IllegalArgumentException()
        }
        val r = FFI.pbmx_bytes_payload(this.handle, buf, buf.remaining().toLong())
        assert(r != 0)
    }

    protected fun finalize() {
        if (!deleted) {
            FFI.pbmx_delete_block_builder(this.handle)
        }
    }
}
